v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/transforms/SourceWriterFn.java

```diff
@@ -0,0 +1,342 @@
+    this.shards = shards;
+    this.spannerConfig = spannerConfig;
+    this.ddl = ddl;
+    this.shadowTablePrefix = shadowTablePrefix;
+  }
+
+  /** Setup function connects to Cloud Spanner. */
+  @Setup
+  public void setup() {
+    mapper = new ObjectMapper();
+    mapper.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);
+
+    // TODO: Support multiple databases
+    for (Shard shard : shards) {
+      String sourceConnectionUrl =
+          "jdbc:mysql://" + shard.getHost() + ":" + shard.getPort() + "/" + shard.getDbName();
+      mySqlDaoMap = new HashMap<>();
+      mySqlDaoMap.put(
+          shard.getLogicalShardId(),
+          new MySqlDao(sourceConnectionUrl, shard.getUserName(), shard.getPassword()));
+    }
+    spannerDao = new SpannerDao(spannerConfig);
+  }
+
+  /** Teardown function disconnects from the Cloud Spanner. */
+  @Teardown
+  public void teardown() throws Exception {
+    if (mySqlDaoMap != null) {
+      for (MySqlDao mySqlDao : mySqlDaoMap.values()) {
+        mySqlDao.cleanup();
+      }
+    }
+    spannerDao.close();
+    mySqlDaoMap.clear();
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) {
+    KV<Long, TrimmedShardedDataChangeRecord> element = c.element();
+    TrimmedShardedDataChangeRecord spannerRec = element.getValue();
+    String shardId = spannerRec.getShard();
+    // Get the latest commit timestamp processed at source
+    try {
+      JsonNode keysJson = mapper.readTree(spannerRec.getMods().get(0).getKeysJson());
+      String tableName = spannerRec.getTableName();
+      com.google.cloud.spanner.Key primaryKey = changeEventToPrimaryKey(tableName, keysJson);
+      String shadowTableName = shadowTablePrefix + tableName;
+      boolean isSourceAhead = false;
+
+      com.google.cloud.Timestamp processedCommitTimestamp =
+          spannerDao.getProcessedCommitTimestamp(shadowTableName, primaryKey);
+      isSourceAhead =
+          processedCommitTimestamp != null
+              && (processedCommitTimestamp.compareTo(spannerRec.getCommitTimestamp()) > 0);
+
+      if (!isSourceAhead) {
+        MySqlDao mySqlDao = mySqlDaoMap.get(shardId);
+
+        InputRecordProcessor.processRecords(
+            new ArrayList<
+                com.google.cloud.teleport.v2.templates.changestream.TrimmedShardedDataChangeRecord>(
+                Arrays.asList(spannerRec)),
+            schema,
+            mySqlDao,
+            shardId,
+            sourceDbTimezoneOffset);
+
+        try {
+          spannerDao.updateProcessedCommitTimestamp(
+              getShadowTableMutation(
+                  tableName, shadowTableName, keysJson, spannerRec.getCommitTimestamp()));
+        } catch (Exception e) {
+          LOG.error("Failed to update last commit timestamp", e);
+          // TODO DLQ handling
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to write to source", e);
+    }
+  }
+
+  private com.google.cloud.spanner.Key changeEventToPrimaryKey(
```

Comment:
I think its better to create this method in spanner-common and use both in forward and reverse replication templates